File: seminar_6/task_1/churilkinartem/task1.cpp
#include <iostream>
#include <vector>
#include <map>

// Функция поиска в глубину
void dfs(const std::vector<std::vector<int>>& graph, int v, std::map<int, bool>& visited, std::vector<int>& component) {
    visited[v] = true;
    component.push_back(v);
    for (int i : graph[v]) {
        if (!visited[i]) {
            dfs(graph, i, visited, component);
        }
    }
}

// Функция поиска связанных компонент в графе
std::vector<std::vector<int>> find_connected_components(const std::vector<std::vector<int>>& graph) {
    std::map<int, bool> visited;
    for (int i = 0; i < graph.size(); ++i) {
        visited[i] = false;
    }

    std::vector<std::vector<int>> connected_components;
    for (int i = 0; i < graph.size(); ++i) {
        if (!visited[i]) {
            std::vector<int> component;
            dfs(graph, i, visited, component);
            connected_components.push_back(component);
        }
    }
    return connected_components;
}

int main() {
    std::vector<std::vector<int>> graph = {
        {1},       // Вершина 0 связана с вершиной 1
        {0, 2},    // Вершина 1 связана с вершинами 0 и 2
        {3},       // Вершина 2 связана с вершиной 3
        {2, 4},    // Вершина 3 связана с вершинами 2 и 4
        {3}        // Вершина 4 связана с вершиной 3
    };

    std::vector<std::vector<int>> components = find_connected_components(graph);

    // Вывод связанных компонент
    for (const auto& component : components) {
        for (int vertex : component) {
            std::cout << vertex << " ";
        }
        std::cout << std::endl;
    }
    return 0;
}

File: seminar_6/task_1/Khamitov.Islam/task_1.py
def find_connected_components(adj_matrix):
    n = len(adj_matrix)
    color = 0
    visited = [color] * n
    components = dict()

    def dfs(vertex, component, color):
        visited[vertex] = color
        component.append(vertex)

        for neighbor in range(n):
            if adj_matrix[vertex][neighbor] == 1 and not visited[neighbor]:
                dfs(neighbor, component, color)

    for vertex in range(n):
        if not visited[vertex]:
            color += 1
            component = []
            dfs(vertex, component, color)
            components[color] = component
            


    return components

# test________________________

matrix = [
    [1, 1, 1, 0, 0, 0, 0, 0],
    [1, 1, 1, 1, 0, 0, 0, 0],
    [1, 1, 1, 1, 0, 0, 0, 0],
    [0, 1, 1, 1, 1, 0, 0, 0],
    [0, 0, 0, 1, 1, 1, 0, 0],
    [0, 0, 0, 0, 1, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 1, 1],
    [0, 0, 0, 0, 0, 0, 1, 1]
]

components = find_connected_components(matrix)
print(components)